src/sentry/snuba/metrics/query.py

```diff
@@ -42,43 +68,160 @@ class MetricsQuery:
     granularity: Granularity
     where: Optional[ConditionGroup] = None  # TODO: Should restrict
     groupby: Optional[Sequence[Groupable]] = None
-    orderby: Optional[OrderBy] = None
+    orderby: Optional[Sequence[OrderBy]] = None
     limit: Optional[Limit] = None
     offset: Optional[Offset] = None
     include_totals: bool = True
     include_series: bool = True
 
-    # TODO: These should be properties of the Histogram field
+    # TODO(ahmed): These should be properties of the Histogram field. We need to extend MetricField
+    #  to accept params and we should pass histogram fields as params on a specific instance of
+    #  MetricField rather than them living on an instance of MetricsQuery
     histogram_buckets: int = 100
     histogram_from: Optional[float] = None
     histogram_to: Optional[float] = None
 
-    def __post_init__(self) -> None:
-        # Validate series limit
-        intervals_len = math.ceil(
-            (self.end - self.start).total_seconds() / self.granularity.granularity
+    @staticmethod
+    def _validate_field(field: MetricField) -> None:
+        derived_metrics_mri = get_derived_metrics(exclude_private=True)
+        metric_mri = get_mri(field.metric_name)
+
+        if field.op:
+            if field.op not in OPERATIONS:
+                raise InvalidParams(
+                    f"Invalid operation '{field.op}'. Must be one of {', '.join(OPERATIONS)}"
+                )
+            if metric_mri in derived_metrics_mri:
+                raise DerivedMetricParseException(
+                    f"Failed to parse {field.op}({field.metric_name}). No operations can be "
+                    f"applied on this field as it is already a derived metric with an "
+                    f"aggregation applied to it."
+                )
+
+    def validate_select(self) -> None:
+        if len(self.select) == 0:
+            raise InvalidParams('Request is missing a "field"')
+        for field in self.select:
+            self._validate_field(field)
+
+    def validate_where(self) -> None:
+        if not self.where:
+            return
+        for condition in self.where:
+            if (
+                isinstance(condition, Condition)
+                and isinstance(condition.lhs, Function)
+                and condition.lhs.function == "ifNull"
+            ):
+                parameter = condition.lhs.parameters[0]
+                if isinstance(parameter, Column) and parameter.name.startswith("tags["):
+                    tag_name = parameter.name.split("tags[")[1].split("]")[0]
```

Comment:
i was wondering if we can improve it or implement in more elegant way.

Reply:
Yeah that's a good point! I was wondering the same actually. I am not super familiar with how DnD folks generate these conditions
I will tag @wmak here to ask if there is a better way to do this check